presto-main/src/main/java/io/prestosql/operator/project/PageProcessor.java

```diff
@@ -200,15 +222,13 @@ private ProjectSelectedPositions(ConnectorSession session, DriverYieldSignal yie
                 verify(result.isSuccess());
                 Page resultPage = result.getPage();
 
-                // if we produced a large page or if the expression is expensive, halve the batch size for the next call
-                long pageSize = resultPage.getSizeInBytes();
-                if (resultPage.getPositionCount() > 1 && (pageSize > MAX_PAGE_SIZE_IN_BYTES || expressionProfiler.isExpressionExpensive())) {
-                    projectBatchSize = projectBatchSize / 2;
+                if (!avoidPageMaterialization) {
+                    updateBatchSize(resultPage.getPositionCount(), resultPage.getSizeInBytes());
                 }
-
-                // if we produced a small page, double the batch size for the next call
-                if (pageSize < MIN_PAGE_SIZE_IN_BYTES && projectBatchSize < MAX_BATCH_SIZE && !expressionProfiler.isExpressionExpensive()) {
-                    projectBatchSize = projectBatchSize * 2;
+                else {
+                    outputPagePositions = resultPage.getPositionCount();
+                    outputPageSizeInBytes = 0;
+                    resultPage = recordMaterializedBytes(resultPage, sizeInBytes -> outputPageSizeInBytes += sizeInBytes);
```

Comment:
There is something pretty subtle going on in this line that I think we need to describe in the code.  I'm thinking about:
* this is passing a mutator for `outputPageSizeInBytes` which is linked into a callback when the blocks load
* what is the thread safety? (I think this is fine due to the restrictions on work processors)
* what happens on a "late" load... say one after the next call to `process`?  Again I think this is ok due to the work processor restrictions